tests/bot/utils/test_redis_cache.py

```diff
@@ -0,0 +1,175 @@
+        self.assertIs(await self.redis.contains('favorite_dentist'), False)
+
+    async def test_items(self):
+        """Test that the RedisDict can be iterated."""
+        await self.redis.clear()
+
+        # Set up our test cases in the Redis cache
+        test_cases = [
+            ('favorite_turtle', 'Donatello'),
+            ('second_favorite_turtle', 'Leonardo'),
+            ('third_favorite_turtle', 'Raphael'),
+        ]
+        for key, value in test_cases:
+            await self.redis.set(key, value)
+
+        # Consume the AsyncIterator into a regular list, easier to compare that way.
+        redis_items = [item async for item in self.redis.items()]
+
+        # These sequences are probably in the same order now, but probably
+        # isn't good enough for tests. Let's not rely on .hgetall always
+        # returning things in sequence, and just sort both lists to be safe.
+        redis_items = sorted(redis_items)
+        test_cases = sorted(test_cases)
+
+        # If these are equal now, everything works fine.
+        self.assertSequenceEqual(test_cases, redis_items)
+
+    async def test_length(self):
+        """Test that we can get the correct .length from the RedisDict."""
+        await self.redis.clear()
+        await self.redis.set('one', 1)
+        await self.redis.set('two', 2)
+        await self.redis.set('three', 3)
+        self.assertEqual(await self.redis.length(), 3)
+
+        await self.redis.set('four', 4)
+        self.assertEqual(await self.redis.length(), 4)
+
+    async def test_to_dict(self):
+        """Test that the .to_dict method returns a workable dictionary copy."""
+        copy = await self.redis.to_dict()
+        local_copy = {key: value async for key, value in self.redis.items()}
+        self.assertIs(type(copy), dict)
+        self.assertDictEqual(copy, local_copy)
+
+    async def test_clear(self):
+        """Test that the .clear method removes the entire hash."""
+        await self.redis.clear()
+        await self.redis.set('teddy', 'with me')
+        await self.redis.set('in my dreams', 'you have a weird hat')
+        self.assertEqual(await self.redis.length(), 2)
+
+        await self.redis.clear()
+        self.assertEqual(await self.redis.length(), 0)
+
+    async def test_pop(self):
+        """Test that we can .pop an item from the RedisDict."""
+        await self.redis.clear()
+        await self.redis.set('john', 'was afraid')
+
+        self.assertEqual(await self.redis.pop('john'), 'was afraid')
+        self.assertEqual(await self.redis.pop('pete', 'breakneck'), 'breakneck')
+        self.assertEqual(await self.redis.length(), 0)
+
+    async def test_update(self):
+        """Test that we can .update the RedisDict with multiple items."""
+        await self.redis.clear()
+        await self.redis.set("reckfried", "lona")
+        await self.redis.set("bel air", "prince")
+        await self.redis.update({
```

Comment:
I don't know if this is a "wrong" way to write unit tests, but it's different from my approach. My problem with doing something like this is that your test not only relies on `update` working, but also `clear` and `set`. If `set` breaks, this test will fail even though it's only supposed to fail if `update` is broken (because it's meant to test `update`). It would be misleading to see this test fail for such reason because my first thought would be something is wrong with `update`, but upon closer inspection, I'd discover it's fine and something else is wrong. I don't know if that's such a bad thing but it's just my take on it.

Reply:
I see your point, but I'm not entirely sure how to test this properly without relying on `set` and `clear` to manage the state of the cache. Do you have a suggestion?

Reply:
One way to look at it is that we assume our redis library is already tested and working. So we can assume that if we use `hmset_dict`, it will work properly. The library already has tests written to ensure this. Thus, all our tests need to do is assert `hmset_dict` was called with the right arguments.

If not that, another way to do it is to use the redis command directly rather than via your class since, again, we assume these to be tested by the library.

On the other hand, these tests could be seen more like integration tests and so they can be left alone? Though the way I've understood integration testing is it tests two separate modules, not just multiple functions in the same module.

I think these tests are still effective and I'm not going to ask you to rewrite them.

Reply:
I appreciate and understand your point. I think I will keep this in mind for future tests, but perhaps leave these to be for the time being, since it would require me to rewrite almost all the tests to achieve this sort of atomic testing at this point.

I think for our purposes, it's probably good enough for now that _some of these tests_ start failing if there's a problem with the RedisCache. But I will try to write more atomic tests in the future.